### PR TITLE
Removed Slurm specific methods from LaunchArguments

### PR DESCRIPTION
**Pull request by juliaputko**

### CODE REVIEW

File: smartsim/_core/arguments/shell.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+from __future__ import annotations
+
+import typing as t
+from abc import abstractmethod
+
+from smartsim.log import get_logger
+from smartsim.settings.arguments.launchArguments import LaunchArguments
+
+if t.TYPE_CHECKING:
+    from smartsim.settings.arguments import LaunchArguments
```
Comment (MattToast) [inline smartsim/_core/arguments/shell.py:36]: Latter import shadows the former. Looks like these are pointing to the same type though, so you should be free to just drop the second:
```suggestion
from smartsim.settings.arguments.launchArguments import LaunchArguments
```

File: smartsim/_core/arguments/shell.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+
+
+class ShellLaunchArguments(LaunchArguments):
+    @abstractmethod
+    def format_env_vars(
+        self, env_vars: t.Dict[str, t.Optional[str]]
+    ) -> t.Union[t.List[str], None]: ...
+
+    @abstractmethod
+    def format_launch_args(self) -> t.Union[t.List[str], None]: ...
```
Comment (MattToast) [inline smartsim/_core/arguments/shell.py:48]: Very minor nit: while we're here can we upgrade to the new style type hints?
```suggestion
    @abstractmethod
    def format_env_vars(
        self, env_vars: dict[str, str | None]
    ) -> list[str] | None: ...

    @abstractmethod
    def format_launch_args(self) -> list[str] | None: ...
```

File: smartsim/_core/arguments/shell.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+if t.TYPE_CHECKING:
+    from smartsim.settings.arguments import LaunchArguments
+
+logger = get_logger(__name__)
+
+
+class ShellLaunchArguments(LaunchArguments):
+    @abstractmethod
+    def format_env_vars(
+        self, env_vars: t.Dict[str, t.Optional[str]]
```
Comment (MattToast) [inline smartsim/_core/arguments/shell.py:44]: pedantic interface question: does `env_vars` here need to be a `dict`? If we are not planning to write into it, could it instead be a `t.Mapping`? The only reason I ask is becuase currently this fails type check:
```py
import os
my_args: ShellLaunchArguments = ...
my_args.format_env_vars(os.environ)
```

File: smartsim/_core/arguments/shell.py (hunk 2 above)
Comment (MattToast) [inline smartsim/_core/arguments/shell.py:48]: Similar pedantic interface question: does this return type need to be nullable? If so, what is the difference between returning `None` and returning `[]`?

**Review by MattToast — CHANGES_REQUESTED** [review]
Looks great! Just a couple of sanity checks!!

File: smartsim/_core/dispatch.py [hunk 4]
```diff
@@ -417,7 +417,7 @@ def make_shell_format_fn(
     """
 
     def impl(
-        args: LaunchArguments, exe: ExecutableProtocol, _env: _EnvironMappingType
+        args: ShellLaunchArguments, exe: ExecutableProtocol, _env: _EnvironMappingType
     ) -> t.Sequence[str]:
         return (
             (
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/_core/dispatch.py]: talked about offline. New ticket created

File: smartsim/settings/launchSettings.py [hunk 5]
```diff
@@ -152,28 +153,28 @@ def update_env(self, env_vars: t.Dict[str, str | None]) -> None:
...
+    #     """
+    #     return self._arguments.format_comma_sep_env_vars(self._env_vars)
+
+    # def format_launch_args(self) -> t.Union[t.List[str], None]:
+    #     """Return formatted launch arguments
+    #     For ``RunSettings``, the run arguments are passed
+    #     literally with no formatting.
+    #     :return: list run arguments for these settings
+    #     """
+    #     return self._arguments.format_launch_args()
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/settings/launchSettings.py]: removed

**Review by MattToast — APPROVED** [review]
A couple of slightly tighter type hints if you want, and some small style failures in the CI, but otherwise looks about ready on my end!

LGTM!!

File: smartsim/settings/arguments/launch/alps.py [hunk 6]
```diff
@@ -176,9 +177,7 @@ def set_quiet_launch(self, quiet: bool) -> None:
         else:
             self._launch_args.pop("quiet", None)
 
-    def format_env_vars(
-        self, env_vars: t.Optional[t.Dict[str, t.Optional[str]]]
-    ) -> t.Union[t.List[str], None]:
+    def format_env_vars(self, env_vars: t.Mapping[str, str | None]) -> list[str] | None:
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/settings/arguments/launch/alps.py:180]: changed

File: smartsim/settings/arguments/launch/local.py [hunk 7]
```diff
@@ -28,27 +28,28 @@
...
+class LocalLaunchArguments(ShellLaunchArguments):
     def launcher_str(self) -> str:
         """Get the string representation of the launcher
 
         :returns: The string representation of the launcher
         """
         return LauncherType.Local.value
 
-    def format_env_vars(self, env_vars: StringArgument) -> t.Union[t.List[str], None]:
+    def format_env_vars(self, env_vars: t.Mapping[str, str | None]) -> list[str] | None:
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/settings/arguments/launch/local.py:52]: changed

File: smartsim/settings/arguments/launch/mpi.py [hunk 8]
```diff
@@ -171,9 +172,7 @@ def set_quiet_launch(self, quiet: bool) -> None:
         else:
             self._launch_args.pop("quiet", None)
 
-    def format_env_vars(
-        self, env_vars: t.Optional[t.Dict[str, t.Optional[str]]]
-    ) -> t.Union[t.List[str], None]:
+    def format_env_vars(self, env_vars: t.Mapping[str, str | None]) -> list[str] | None:
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/settings/arguments/launch/mpi.py:175]: changed

File: smartsim/settings/arguments/launch/pals.py [hunk 9]
```diff
@@ -105,9 +106,7 @@ def set_hostlist(self, host_list: t.Union[str, t.List[str]]) -> None:
             raise TypeError("host_list argument must be list of strings")
         self.set("hosts", ",".join(host_list))
 
-    def format_env_vars(
-        self, env_vars: t.Optional[t.Dict[str, t.Optional[str]]]
-    ) -> t.Union[t.List[str], None]:
+    def format_env_vars(self, env_vars: t.Mapping[str, str | None]) -> list[str] | None:
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/settings/arguments/launch/pals.py:109]: changed

File: smartsim/settings/arguments/launch/slurm.py [hunk 10]
```diff
@@ -251,9 +252,7 @@ def format_launch_args(self) -> t.Union[t.List[str], None]:
                     formatted += ["=".join((prefix + key, str(value)))]
         return formatted
 
-    def format_env_vars(
-        self, env_vars: t.Dict[str, t.Optional[str]]
-    ) -> t.Union[t.List[str], None]:
+    def format_env_vars(self, env_vars: t.Mapping[str, str | None]) -> list[str] | None:
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/settings/arguments/launch/slurm.py:255]: changed

File: smartsim/_core/arguments/shell.py [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+logger = get_logger(__name__)
+
+
+class ShellLaunchArguments(LaunchArguments):
+    @abstractmethod
+    def format_env_vars(
+        self, env_vars: t.Mapping[str, str | None]
+    ) -> list[str] | None: ...
+    @abstractmethod
+    def format_launch_args(self) -> list[str] | None: ...
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/_core/arguments/shell.py:44]: changed

File: smartsim/settings/launchSettings.py [hunk 12]
```diff
@@ -29,6 +29,7 @@
 import copy
 import typing as t
 
+from smartsim._core.arguments.shell import ShellLaunchArguments
```
[the remark this answers is not in the document]
Comment (juliaputko) [reply smartsim/settings/launchSettings.py]: removed

**Pull request merged**